Adds ability to smooth lines for line and area charts

stormpython opened the pull request:
Closes #825.

Adds ability to smooth lines in line chart.

rashidkpc commented:
Can you also add this to the area chart? There's no interpolate() call in there at the moment, but I gave it a shot and it seemed to work swimmingly.

stormpython commented:
@rashidkpc done.

w33ble commented:
Seems like the bound padding might need to be increased when using smooth interpolation, as the lines can extend outside the bounds of the view.

![screenshot 2015-03-06 11 17 13](https://cloud.githubusercontent.com/assets/404731/6530952/6746e534-c3f2-11e4-8cc7-1de67f2cb83c.png)

w33ble commented:
Can you fix the UI a little while you're in here too? Specifically, just add the `form-control` class to the mode select, so they match...

![screenshot 2015-03-06 11 47 19](https://cloud.githubusercontent.com/assets/404731/6531415/958ffc06-c3f6-11e4-8da3-a189b73fbf2d.png)

stormpython commented:
Increased the line tension and added the `form-control` class.

rashidkpc commented on src/kibana/plugins/vis_types/controls/line_interpolation_option.html:
```diff
@@ -0,0 +1,7 @@
+<div>
+  <label>
+    Line Interpolation
```
Can we `s/Interpolation/Smoothing` this? I don't think most people know what interpolation is. Also, if there's only on or off (linear or smooth) and we don't anticipate adding more options than that why not just make it a checkbox?

rashidkpc commented:
A bit wacky on percentage charts:

![image](https://cloud.githubusercontent.com/assets/1250387/6546590/df400cf2-c576-11e4-9c18-523ce4cede47.png)

stormpython commented:
Renamed `Line Interpolation` to `Line Smoothing`. Removed the upper buffer from the svg `clipPath` on area charts since it was not needed. This solves the issue of lines in area charts going off the y axis.

w33ble commented:
I'm curious if there are other ways to do line interpolation besides smooth and linear. My thinking being that if we're going to add more, that label should be a little less specific.

w33ble commented:
Still seeing line clipping with smooth interpolation

![screenshot 2015-03-17 11 26 04](https://cloud.githubusercontent.com/assets/404731/6694570/54510d70-cc98-11e4-9dba-5f5d9e4f2381.png)

stormpython commented:
@w33ble try this out and see if it fixes the issue.

w33ble commented:
The lines still clip in the negative axis

![screenshot 2015-03-19 11 09 09](https://cloud.githubusercontent.com/assets/404731/6737347/48c012f8-ce28-11e4-897e-1fe1816aba2b.png)

I'm also not sure about this fix - just adding an extra buffer of 1 works in my case, but if the data set spans a large range, 1 unit on the y-axis may not be enough. Is there not a way to get the bounds of the lines that are generated and ensure they stay in the view?

stormpython commented:
@w33ble An algorithm calculates the curvature of the line. If we wanted to know how curvy the lines would be, we'd need to do some measurement at every possible point on the equation that is used to calculate the position of each point on a line. Can you show me an example where increasing the y max value by 1 still shows lines being cut-off? I will decrease the yMin values when using the data extents by 1 to make sure this doesn't happen on the lower end.

If you can find or think of a better solution, let me know. As of now, this seems as good a fix as any.

stormpython commented:
Ok, it would appear that the solution to the lines going off the chart is not to use the `cardinal` option for smoothing, but use the `monotone` option which does all the smoothing but without the added fact of lines overshooting the axis.

rashidkpc commented:
@stormpython Can you merge master on this one?

w33ble commented:
The new smoothing tends to leave the border of the vis less, but it still happens (not the bottom of the chart).

![image](https://cloud.githubusercontent.com/assets/404731/6975420/de485492-d950-11e4-94c3-ddd4e9006f7d.png)

The smoothing is less "smooth" now as a result though (note the very pointed peaks here). Perhaps this could perhaps be solved by offering users the ability to pick their smoothing method?

@rashidkpc how much do we care that the lines leave the border of the svg? It seems to happen regardless, and seems like a tricky problem to solve - it's the number one reason this PR keeps getting sent back.

stormpython commented:
@w33ble the lines aren't actually leaving the bottom of the chart there. It only appears to since there are points there with zero values. It is simply going through the point of the circle. The reason it appears cut off is because part of the circle is cut-off, and that is due to the clip path that is set. Half the circle is above the 0 line and half is below. I don't really consider this to be a bad thing. 

As far as the smoothness goes, I can reduce the tension on the line to make it more curvy.

rashidkpc commented:
I personally don't much care that they leave the plot. If you choose to have smooth lines you choose to a lie :-)

stormpython commented:
So, reducing the tension doesn't actually change the slope by much of the line between points. So I could either leave monotone the way it is or use cardinal (and have times where the lines go off the scale).

w33ble commented:
I liked how cardinal looked a lot better. I think most datasets probably look bad with monotone, but I'm open to letting the user choose (but defaulting to cardinal).

stormpython commented:
@w33ble after talking with @rashidkpc, he agreed with you that `cardinal` is a much better line interpolation algorithm than `monotone`. I've changed the drop down option to a checkbox option and sticking with `cardinal`. Since this pull should be getting in with #3464, users will have the ability to change the min and max y axis values if parts of lines get clipped.
